### PR TITLE
[truncate explanation] Fix edge case when we truncate due to max_chars

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
Previousely the added test case would output:
"...Full output truncated (0 lines hidden), use '-vv' to show"

Also removed some useless complexity.

### CODE REVIEW

File: src/_pytest/assertion/truncate.py [hunk 1]
```diff
@@ -88,29 +88,23 @@ def _truncate_explanation(
         truncated_explanation = input_lines[:max_lines]
     else:
         truncated_explanation = input_lines
-    truncated_char = True
     # We reevaluate the need to truncate chars following removal of some lines
-    if len("".join(truncated_explanation)) > tolerable_max_chars and max_chars > 0:
+    need_to_truncate_char = (
```
Comment (Pierre-Sassoulas) [inline src/_pytest/assertion/truncate.py:92]:
```suggestion
    need_to_truncate_char = (
        # '70 < tolerable_max_chars' means 'max_chars > 0'
```

Comment (bluetech) [inline src/_pytest/assertion/truncate.py:92]: If I understand correctly (tell me if I'm wrong), `max_lines == 0` means "don't truncate lines" and `max_chars == 0` means "don't truncate chars", right? I think then it would be more obvious to not do `70 < tolerable_max_chars` and instead do `max_chars > 0` for the "is chars truncation enabled?" check like before. It's equivalent but fits better with the `if max_lines > 0` check a few lines above it.

It will also be nice to document the meaning of `0`s in the docstring of the function, just so local reasoning about it is easier.

File: src/_pytest/assertion/truncate.py [hunk 2]
```diff
@@ -88,29 +88,23 @@ def _truncate_explanation(
...
             truncated_explanation, max_chars
         )
-    else:
-        truncated_char = False
 
     if truncated_explanation == input_lines:
         # No truncation happened, so we do not need to add any explanations
         return truncated_explanation
 
-    truncated_line_count = len(input_lines) - len(truncated_explanation)
```
Comment (bluetech) [inline src/_pytest/assertion/truncate.py:104]: Am I correct that a minimal fix for the issue would be

```diff
diff --git i/src/_pytest/assertion/truncate.py w/src/_pytest/assertion/truncate.py
index 5820e6e8a..fbe18fd3d 100644
--- i/src/_pytest/assertion/truncate.py
+++ w/src/_pytest/assertion/truncate.py
@@ -105,12 +105,12 @@ def _truncate_explanation(
     if truncated_explanation[-1]:
         # Add ellipsis and take into account part-truncated final line
         truncated_explanation[-1] = truncated_explanation[-1] + "..."
-        if truncated_char:
-            # It's possible that we did not remove any char from this line
-            truncated_line_count += 1
     else:
         # Add proper ellipsis when we were able to fit a full line exactly
         truncated_explanation[-1] = "..."
+    if truncated_char:
+        # It's possible that we did not remove any char from this line
+        truncated_line_count += 1
     return [
         *truncated_explanation,
         "",
```

If so, I think it would be good to do the minimal fix in its own commit, just to separate the logic fix from the refactoring (which should have no logical changes), that's a bit easier to review.

**Review by bluetech — COMMENTED** [review]
Thanks! Please see my comments.

File: src/_pytest/assertion/truncate.py (hunk 2 above)
Comment (Pierre-Sassoulas) [reply src/_pytest/assertion/truncate.py:104]: Yeap, sorry about the bad commit history, I fixed that.

File: src/_pytest/assertion/truncate.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply src/_pytest/assertion/truncate.py:92]: Indeed, it felt close to not worth it while writing it 😄

**Comment by Pierre-Sassoulas** [comment]
Drafting because I realized that there's a lot of optimization that can still be done.
